crossplane-capi-import/README.md

```diff
@@ -0,0 +1,315 @@
+
+During the generation process, there are certain details that may (excluding the
+use of composition functions) require the editing of specific blocks of the
+composition.
+
+The reason for this is that there is no looping inside the composition and
+details returned from the cloud do not directly correlate to information required
+by CAPI.
+
+The example of this inside EKS is the subnet groups required by the
+`AWSManagedControlPlane` resource which then require hard coded specification
+via patching.
+
+```yaml
+- fromFieldPath: status.subnetIds[0]
+  toFieldPath: spec.forProvider.manifest.spec.network.subnets[0].id
+- fromFieldPath: status.subnetIds[1]
+  toFieldPath: spec.forProvider.manifest.spec.network.subnets[1].id
+- fromFieldPath: status.subnetIds[2]
+  toFieldPath: spec.forProvider.manifest.spec.network.subnets[2].id
+- fromFieldPath: status.subnetIds[3]
+  toFieldPath: spec.forProvider.manifest.spec.network.subnets[3].id
+- fromFieldPath: status.subnetIds[4]
+  toFieldPath: spec.forProvider.manifest.spec.network.subnets[4].id
+- fromFieldPath: status.subnetIds[5]
+  toFieldPath: spec.forProvider.manifest.spec.network.subnets[5].id
+```
+
+This issue can to a large degree be solved through the use of composition
+functions and to support this, there is a proof of concept on how this can work
+placed at [https://github.com/giantswarm/crossplane-fn-generate-subnets/](https://github.com/giantswarm/crossplane-fn-generate-subnets/).
+
+To provide 100% accurate results for CAPA, not all fields can be recovered
+directly through composition functions, in particular the capability of
+identifying public vs private subnets.
+
+This would be solvable in future iterations as the upstream issue on [Querying
+and filtering for import and observe](https://github.com/crossplane/crossplane/issues/4141)
+is resolved as it would enable the discovery of route tables in the same manner
+CAPA tracks this today.
+
+Within the PoC, I "fake" this by proposing either a custom tag be added to the
+subnet, or by looking to see if `mapPublicIpOnLaunch` is set to true on the
+subnet, although I dislike the second option as not every public subnet would
+have this flag set. For the first option, the PoC offers that a tag of
+`giantswarm.io/public` be set although this is open to refinement.
+
+### Clusters with multiple nodepools
+
+Crossplane requires a direct one to one mapping between Custom Resource and
+Cloud Infrastructure types. This makes it more difficult to directly reconcile
+multiple nodepools without building custom compositions for each variant of
+customer infrastructure we may encounter.
+
+There are two potential solutions to this problem:
+
+1. We develop a library of common structures that we encounter. This will be
+   built as part of discovery with a custom using boilerplate code to  enhance
+   the compositions for each customer.
```

Comment:
But will a logic needed to map these structures to our library entries be manageable? Won't it become a pile of huge `if`s?

Reply:
In this instance no. The idea of a library here would have been that we develop an XRC for each cluster configuration we found. E.g. one with 3 subnets defiend, one with 6, etc.

In my mind this was never really a viable option as it requires constant maintenance.